### PR TITLE
Implement maxout layer in mlpack.

### PR DESCRIPTION
**Pull request by prasanna08**
This PR adds maxout layer in mlpack ann module.

### CODE REVIEW

File: src/mlpack/methods/ann/layer/maxout_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  bias = arma::mat(weights.memptr() + weight.n_elem,
+      outSize * hiddenSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (hiddenParameter.n_elem == 0) {
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Minor style issue please put the opening bracket on a new line; For more information: https://github.com/mlpack/mlpack/wiki/DesignGuidelines#brace-placement

I tend to be very  picky about code but I am not mean :)

File: src/mlpack/methods/ann/layer/maxout_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    hiddenParameter = arma::zeros<arma::mat>(outSize * hiddenSize,
+        input.n_cols);
+  }
+
+  if (output.n_elem == 0) {
+    output = arma::zeros<arma::mat>(outSize, input.n_cols);
+  }
+
+  hiddenParameter = (weight * input) + bias;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Do we have to initialize the hiddenParameter in line 57 if we just assign the parameter here?

File: src/mlpack/methods/ann/layer/maxout_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  for (int i = 0, j = 0; j < outSize; i+=hiddenSize, j++) {
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: We usually use size_t as index type, also another picky comment can you put a space between i and i+= hiddenSize.

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 4]
```diff
@@ -410,4 +410,134 @@ BOOST_AUTO_TEST_CASE(DropConnectNetworkTest)
...
+  for (size_t i = 0; i < dataset.n_cols; ++i)
+    dataset.col(i) /= norm(dataset.col(i), 2);
+
+  arma::mat labels = arma::zeros(1, dataset.n_cols);
+  labels.submat(0, labels.n_cols / 2, 0, labels.n_cols - 1).fill(1);
+  labels += 1;
+
+  // Maxout neural net with logistic activation function.
+  BuildMaxoutNetwork<>
+      (dataset, labels, dataset, labels, 2, 5, 10, 50, 0.2);
```
Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp]: I didn't expected an error of 0.2 using 50 iterations, nice.

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 5]
```diff
@@ -410,4 +410,134 @@ BOOST_AUTO_TEST_CASE(DropConnectNetworkTest)
...
+                        const size_t maxEpochs,
+                        const double classificationErrorThreshold)
+{
+ /*
+  *  Construct a feed forward network with trainData.n_rows input nodes,
+  *  maxoutUnitSize hidden nodes and trainLabels.n_rows output nodes. The
+  *  network struct that looks like:
+  *
+  *  Input         Maxout         Output
+  *  Layer         Layer          Layer
```
Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp]: Shouldn't this be input -> maxout -> linear -> output?

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 2 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: No that is not necessary. We can do direct assignment.

File: src/mlpack/tests/feedforward_network_test.cpp (hunk 5 above)
Comment (prasanna08) [inline src/mlpack/tests/feedforward_network_test.cpp]: No. Actually I transfer output from maxout layer to a linear layer and this linear layer's output is output probability (on which log-likelihood with softmax is applied). So the final linear layer is actually output layer.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 1 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Sure. Design guidelines are important. Otherwise code looks messy.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+void Maxout<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.n_elem == 0)
+    output = arma::zeros<arma::mat>(outSize, input.n_cols);
+
+  hiddenParameter = (weight * input) + bias;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+    output.row(j) = arma::max(hiddenParameter.rows(i, i + hiddenSize - 1), 0);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Do you think using submat (arma::mat(hiddenParameter.submat(), 0) is faster as iterating over the rows?

File: src/mlpack/methods/ann/layer/maxout_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      hiddenParameter.n_cols);
+  arma::mat hiddenParameterSubset;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+  {
+    hiddenParameterSubset = hiddenParameter.rows(i, i + hiddenSize - 1);
+    
+    for (int k = 0; k < hiddenParameterSubset.n_cols; k++)
+    {
+      int ind = arma::as_scalar(arma::find(arma::max(
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: We should specify the number of indices find returns.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 6 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: I can't say. I don't have much experience with arma. If you think that's better way then I will change implementation.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 7 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: It returns only one index. Basically what I am trying is finding the hidden unit that had maximum activation because the error should be backpropagated to this hidden unit only. So I am finding index of this maximum activation in for all test cases and backpropagating error for them.

I will add comment describing this. Also would you confirm about this backpropagation? I am not sure whether I wrote it correctly or not.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 6 above)
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: I'm not sure, it's faster, do you think we can write a simple benchmark to test that out? Something like that:

```
arma::mat input = arma::randu(20000, 1);
arma::mat output;

auto start = chrono::steady_clock::now();

layer.Forward(std::move(input), std::move(output));

auto end = chrono::steady_clock::now();
auto diff = end - start;
cout << "maxout: " << chrono::duration <double, milli> (diff).count() << " ms" << endl;
```

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 7 above)
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: It looks right, it's basically what you do for MaxPooling, where I also used the same approach:

```
class MaxPoolingRule
{
 public:
  /*
   * Return the maximum value within the receptive block.
   *
   * @param input Input used to perform the pooling operation.
   */
  template<typename MatType>
  size_t Pooling(const MatType& input)
  {
    return arma::as_scalar(arma::find(input.max() == input, 1));
  }
};
```

But I think I can construct a case where max isn't just a single element e.g. ``arma::max({0, 0, 0})`` or ``arma::max({0.5, 0.5, 0})`` here max returns more than one element, which isn't a problem, we can just use the first max. Besides, specifying the number would probably result in a speedup since you can break once you found the first occurrences of the maximum. Let me know what you think.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 6 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Yeah sure. I will test it and report back.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 7 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Yup. You are correct. Specifying number directly would result in speedup. Also, here we don't actually want the maximum value. The point of storing hiddenParameter is to get index of max element at the time of backprop and gradient. I think we can optimize this by directly storing indices of max element during forward pass (where we have to find max element for output). Then later in backprop or gradient we can directly utilize these indices. What do you think ?

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 6 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: I tested and both are performing equally well.
Data I observed:

<table>
<tr>
<th>method</th>
<th>in_size</th>
<th>hidden_size</th>
<th>out_size</th>
<th>run</th>
<th>time</th>
</tr>
<tr>
<td>rows</td>
<td>20000</td>
<td>7</td>
<td>5</td>
<td>1st</td>
<td>1.54378 ms</td>
</tr>
</tr>

<tr>
<td>rows</td>
<td>20000</td>
<td>7</td>
<td>5</td>
<td>2nd</td>
<td>0.783643 ms</td>
</tr>


<tr>
<td>submat</td>
<td>20000</td>
<td>7</td>
<td>5</td>
<td>1st</td>
<td>0.772132 ms</td>
</tr>

<tr>
<td>submat</td>
<td>20000</td>
<td>7</td>
<td>5</td>
<td>2nd</td>
<td>0.884035 ms</td>
</tr>

<tr>
<td>rows</td>
<td>20000</td>
<td>100</td>
<td>20</td>
<td>1st</td>
<td>33.3094 ms</td>
</tr>

<tr>
<td>rows</td>
<td>20000</td>
<td>100</td>
<td>20</td>
<td>2nd</td>
<td>33.1371 ms</td>
</tr>

<tr>
<td>submat</td>
<td>20000</td>
<td>100</td>
<td>20</td>
<td>1st</td>
<td>36.081 ms</td>
</tr>

<tr>
<td>submat</td>
<td>20000</td>
<td>100</td>
<td>20</td>
<td>2nd</td>
<td>32.0741 ms</td>
</tr>
</table>

So according to my observation its a tie. So its your call.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 7 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Hi @zoq I bench marked the maxout code with 3 different implementations which are:
**maxoutIndices**: Storing index of maximum hidden units and then using it.
**maxoutValues**: Storing value of maximum hidden unit and then using it (basically storing output of maxout layer).
**hiddenValues**: current approach I have implemented.

I tested it by training end to end model for all 3 setups (since this changes affected all forward backward and gradient passes). 

<table>
<tr>
<th>method</th>
<th>iterations</th>
<th>time</th>
</tr>

<tr>
<td>maxoutIndices</td>
<td>70</td>
<td>55100 ms </td>
</tr>
<tr>
<td>maxoutIndices</td>
<td>7</td>
<td>4552.45 ms</td>
</tr>
<tr>
<td>maxoutValues</td>
<td>70</td>
<td>57945.5 ms</td>
</tr>
<tr>
<td>maxoutValues</td>
<td>7</td>
<td>4346.81 ms</td>
</tr>
<tr>
<td>hiddenParameters</td>
<td>70</td>
<td>52324.9 ms</td>
</tr>
<tr>
<td>hiddenParameters</td>
<td>7</td>
<td>4373.22 ms</td>
</tr>
</table>

Honestly I expected slight performance boost from maxoutIndices and maxoutValues method but from observations it seems that hiddenParameters outperforms both of them. I am confused by this result. So I leave this decision on you, too.

Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: I also thought that reusing the indices from the Forward pass would result in an speedup, can you share the code for the other two options you tested out?

Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Btw. I like the way you approach this, it's easier to make a decision based on real numbers.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 6 above)
Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: I agree, I think in that case we can go with the solution you like.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 7 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: I used following code in **maxoutIndices**

in forward pass I used following loop.
```
for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++) {
    hiddenParameterSubset = hiddenParameter.submat(i, 0, i + hiddenSize - 1, 0);
    output.row(j) = arma::max(hiddenParameterSubset, 0);

    for (int k = 0; k < output.n_cols; k++) {
      maxoutIndices(j, k) = arma::as_scalar(
          arma::find(hiddenParameterSubset == output(j, k)));
    }
```

Calculating index in backward and gradient method.
```
int ind = maxoutIndices(j, k);
```

Similarly in **maxoutValue**

forward:
```
for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++) {
    hiddenParameterSubset = hiddenParameter.submat(i, 0, i + hiddenSize - 1, 0);
    output.row(j) = arma::max(hiddenParameterSubset, 0);
    maxoutValues.row(j) = arma::max(hiddenParameterSubset, 0);
}
```

backward and gradient:
```
int ind = arma::as_scalar(arma::find(maxoutValues(j, k) == hiddenParameterSubset.col(k)));
```

Well initially I was doing timing calculations to measure how much boost I am able to get from these methods but then I did not get any improvement so I ran tests again and again for 3 times and I still got same results. Therefore I posted them here because I am confused from these results.

Also I used thyroid dataset for bench marking. I don't think that these test results are influenced by dataset, are they ?

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 6 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: I would go with submat solution then.

File: src/mlpack/methods/ann/layer/maxout_impl.hpp (hunk 7 above)
Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: So basically **maxoutIndices** and **maxoutValues** are similar except where index value is getting calculated. maxoutIndices calculates indices in forward pass (and stores it) and maxoutValues calculates them in Backward and Gradient pass (latter calculates twice but former calculates for once).

Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Hi @zoq, have you decided which implementation to go with? Its the final remaining thing so that this PR can be merged.

Thanks

Comment (zoq) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: Sorry for the slow response, I would go with calculating the maxoutIndices values once in the Forward.

Comment (prasanna08) [inline src/mlpack/methods/ann/layer/maxout_impl.hpp]: No worries. I understand that you have a buys schedule. I will push new commit soon.

Thanks

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 8]
```diff
@@ -410,4 +410,134 @@ BOOST_AUTO_TEST_CASE(DropConnectNetworkTest)
...
+  }
+
+  double classificationError = 1 - double(error) / testData.n_cols;
+  BOOST_REQUIRE_LE(classificationError, classificationErrorThreshold);
+}
+
+/**
+ * Train the maxout network on a larger dataset.
+ */
+BOOST_AUTO_TEST_CASE(MaxoutNetworkTest)
```
Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp]: Can you tell us something about the runtime of this test? ``time bin/mlpack_test -t FeedForwardNetworkTest/MaxoutNetworkTest``

Comment (prasanna08) [inline src/mlpack/tests/feedforward_network_test.cpp]: The output of time was following:
```
real	0m45.613s
user	0m54.128s
sys	1m57.040s
```

Also I was wondering if there is a way to run only selected tests. Thanks for the info.

Comment (prasanna08) [inline src/mlpack/tests/feedforward_network_test.cpp]: Hi @zoq, is there anything remaining here?

Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp]: Do you think we could slightly decrease the number of training iterations?

Comment (prasanna08) [inline src/mlpack/tests/feedforward_network_test.cpp]: Hi @zoq I decresed number of iterations by fraction of 10 (7 and 5 respectively) and test is successful with less iterations. Here is output of `time`.
```
real	0m10.143s
user	0m11.111s
sys	0m24.168s
```

Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp]: That looks great, can you push the changes?

Comment (prasanna08) [inline src/mlpack/tests/feedforward_network_test.cpp]: @zoq done.

**Comment by zoq** [comment]
I'd like to test the layer with a couple models before I'll merge it in, but besides that it looks good. I'll see if I can test the layer with the models in the next days.
